adding doc and diagrams for composition (#1960)
risc0/risc0: website/api/zkvm/composition.md

# Proof Composition

> RISC Zero's zkVM is the only zkVM that offers a convenient API for efficiently verifying proofs inside the guest program.

Reviewer: This is somewhat of a tenuous statement, as SP1 provide a reasonable API and I'm sure others will in the future. Maybe we just talk about how great composition is without comparing it to others.

https://github.com/succinctlabs/sp1/blob/main/examples/aggregation/script/src/main.rs

> RISC Zero's zkVM is the only zkVM that offers a convenient API for efficiently verifying proofs inside the guest program.

Reviewer 2: ```suggestion
RISC Zero's zkVM offers a convenient API for efficiently verifying proofs inside the guest program.
```

This feature enables rich possibilities for building modular applications. Check out the [blog post] to learn more about the power of proof composition.

## Start Building

> To use proof composition in your program, you'll need to call [`add_assumption()`] on the host-side and [`env::verify()`] on the guest-side, as demonstrated in this [code example].

Reviewer: The high level idea of assumptions and resolution could definitely use an introduction first. It might useful to introduce the hypothetical idea of actually just running the verifier code in the zkVM, which is what we tried first, and then explain how the assumptions mechanism improves upon that. That's kind of what I tried to do in my talk for the study club.

Reviewer 2: https://www.youtube.com/watch?v=x0-7Y46bQO0 I just watched an needed to re-watch the assumptions bit 5 times to get it to sink in :sweat_smile: . I had the "naive" model of executing the verifier in another guest, the recursion stuff was new to me with no other exposure to the concept prior

## How it works

Under the hood, proof composition works by _adding assumptions_ to the [ReceiptClaim] struct, and then _resolving_ those assumptions.

![ReceiptClaim with no assumptions][composition-no-assumptions]

### Adding assumptions

When [`env::verify()`] is called inside the guest program, an [assumption] is added to the [ReceiptClaim].
This results in a "conditional receipt."

![Adding an assumption][composition-add-assumption]

### Resolve an assumption

In order to finish the process of proof composition, assumptions must be resolved.

> This is accomplished via [`resolve()`], which is called automatically when users [`verify()`][TODO - is this right?] a conditional receipt.

Reviewer 2: blocking on this

# Proof Composition

RISC Zero's zkVM offers a convenient API for efficiently verifying proofs inside the guest program.

This feature enables rich possibilities for building modular applications. Check out the [blog post] to learn more about the power of proof composition.

:::info

> Although the "obvious" approach to proof composition would be to simply run the verifier inside the guest program, this approach is very expensive.

Reviewer: It's actually not so expensive that there are not cases where folks will use this approach. In particular, there is https://github.com/risc0/risc0/issues/1355. The real core reason is you don't get compression because verifying two receipts takes more than two segments, and so generates more than two receipts. You need verification of two receipts to generate one receipt.
```suggestion
Although the "obvious" approach to proof composition would be to simply run the verifier inside the guest program, this approach doesn't result in compression.
```

For a more detailed dive into proof composition and RISC Zero's approach to recursive proving, check out the [Study Club recording].

[`add_assumption()`]: https://github.com/risc0/risc0/blob/v1.0.1/examples/composition/src/main.rs#L29
[assumption]: /terminology#assumption

> [assumptions]: /terminology#assumption

Reviewer 3: This could be collapsed into a single reference link.

Author: I'm struggling to remember/guess/google the appropriate syntax...

Reviewer 3: `[link-id][link-text]`

Reviewer 3: https://riptutorial.com/markdown/example/2215/reference-link

Reviewer 2: https://github.com/risc0/risc0/pull/1960/commits/33d26bc0dc920af1264e6ea6101dd9033ce00b5f

Author: > [link-id][link-text]
ahh yes ok. 

> https://github.com/risc0/risc0/commit/33d26bc0dc920af1264e6ea6101dd9033ce00b5f
Thanks Nuke!
